libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Extensions/RestClient.swift

```diff
@@ -0,0 +1,432 @@
+ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+import Combine
+/// Errors that can be thrown while using RestClient
+public enum RestClientError: Error {
+    case ApiResponseIsEmpty
+    case ApiInvocationFailed(underlyingError: Error, urlResponse: URLResponse?)
+    case DecodingFailed(underlyingError: Error)
+    case JsonSerialization(underlyingError: Error)
+}
+
+public struct RestResponse {
+    
+    private static let emptyJsonDictionaryArrayResponse = [String:Any]()
+    private static let emptyJsonArrayResponse = [[String:Any]]()
+    private static let emptyStringResponse = ""
+    
+    private (set) var data: Data
+    private (set) var urlResponse: URLResponse
+    
+    /// Initializes the RestResponse with a Data object and URLResponse
+    /// - Parameter data: Response as raw Data
+    /// - Parameter urlResponse: URlResponse from endpoint
+    public init(data: Data, urlResponse: URLResponse) {
+        self.data = data
+        self.urlResponse = urlResponse
+    }
+    
+    /// Parse the response as a Json Dictionary.
+    public func asJson() throws -> Any {
+        do {
+            let jsonData = try JSONSerialization.jsonObject(with: data, options: [])
+            return jsonData
+        } catch let error {
+            throw RestClientError.JsonSerialization(underlyingError: error)
+        }
+    }
+    
+    /// Get response as Data object. Use this for retrieving  binary objects
+    public func asData() -> Data {
+        return self.data
+    }
+    
+    /// Parse response as String
+    public func asString() -> String {
+        let stringData = String(data: data, encoding: String.Encoding.utf8)
+        return stringData ?? RestResponse.emptyStringResponse
+    }
+    
+    /// Decode the response as  a codable
+    /// - Parameter type: The type to use for decoding.
+    public func asDecodable<T:Decodable>(type: T.Type) throws -> Decodable?  {
+        let decoder = JSONDecoder()
+        do{
+            let object = try decoder.decode(type, from: data)
+            return object
+        } catch let error {
+            throw RestClientError.DecodingFailed(underlyingError: error)
+        }
+    }
+}
+
+extension RestClient {
+    
+    /// Execute a request.
+    /// - Parameter request: RestRequest object
+    /// - Parameter completionBlock: The completion block to invoke.
+    public func send(request: RestRequest, _ completionBlock: @escaping (Result<RestResponse,RestClientError>) -> () ) {
```

Comment:
I wonder if we should just provide this one (and the iOS13 publisher one) and let them use the RestRequest factory methods.

Reply:
Maybe the other methods (query/queryAll/search/describe etc etc) are only useful if they return a better typed response. When we know the request, we know the shape of the response.

Reply:
@wmathurin  I agree if we decide to internalize the other apis. i.e. these apis should not  be available in swift. Otherwise it will just add to the confusion (combine, resulttype &  older api shape)